cores/esp32/IPAddress.cpp

```diff
@@ -336,6 +210,12 @@ bool IPAddress::fromString6(const char *address) {
             colons++;
             acc = 0;
         }
+        else if (c == '%') {
+            _zone = netif_name_to_index(address);
```

Comment:
Is this dependency on netif necessary, or cause difficulties. e.g. can it be unit tested?

We also aren't checking the return value, e.g. what if the return is NETIF_NO_INDEX ... should the parse return false?  Sure "fe80::1:2:3:4%eth0" might be expected, by what about "fe80::1:2:3:4%nosuchinterfaceexists"

Also, does "fe80::1:2:3:4%1" parse correctly (returning 1), because usually you expect toString() and fromString() to round-trip. (I don't know... can a unit test validate?)

Or is netif_name_to_index only for names? i.e. if "zone" is numeric then use the numeric value (i.e. atoi()) ourselves, and we only call netif if it is not numeric?

Reply:
the dependency is necessary, because that is how LwIP on the ESP32 works. We want to convert back and fourth with the correct values. As stated above, input fuzzing will come in the future. For now we want to stay as close as possible to official code and add the minimal we need to start supporting it all. After all looks good and seems working, we can delve into securing the code against garbage.

Reply:
Okay, I guess invalid things will just not work anyway.

I probably disagree on how LwIP works as well ... LwIP takes the number zone identifier as the actual value , i.e. _zone ... so fe80::1:2:3:4%1 is a valid IPv6 scoped address, and "1" is a valid identifier. That is the minimum of what is needed.

Similarly, to convert an address structure to a string, we output zone_id directly (if not 0), i.e. as "1".

The conversion from "1" to "eth0" (interface name) is an optional thing on top of that, because it may be easier to recognise than trying to look up the interface numbers. Same for the parsing. Parsing fe80::1:2:3:4%1 should be mandatory; parsing fe80::1:2:3:4%eth0 is an extra bonus.

=> So long as fe80::1:2:3:4%1 works, I would be happy.

Reply:
So, for printTo()/toString() we output the more human-readable netif name.

For parsing if we can support both that would be good. 

netif_name_to_index looks like it only translates (and expects a name), so I think we can do something like atoi() to see if it parses as an number, and if not (result is 0) call netif_name_to_index. Or do the reverse, i.e. check the name first and if that fails (NETIF_NO_INDEX) try atoi().

Reply:
Wouldn't it make sense to have some default (enum like) enumeration where we just have some fixed values based on the type of interface when it refers to some named interface?
On Linux this network-adapter-naming also has become quite a mess, where the order in which devices are detected depends on the kernel you boot. 

I just want to make sure that the numbers remain the same whether you boot a sketch with ETH included or not.
Just so the configuration on the same device doesn't change.
Or has the IDF or LWIP implementation already something similar?

Reply:
Zone ID (the numeric value) is the base, and what you internally send to the network stack. There are limits to making it stable, e.g. changing networks, plugging into docking stations / dongles, setting up VPN tunnels, software networks (like to hosted VM/container networks), etc.

You can't even necessarily detect if two interfaces are actually connected to the same link (i.e. same zone). RFC 4007 has an example of this.

Hopefully on a relatively simple device, with limited interfaces, you will get relatively stable numbering and naming.